File: Wells/Source/Filters/StringFilter.vb
Imports Wells.CorePersistence.Repositories

Public Class StringFilter(Of T)
    Inherits BaseFilter(Of T)

    Property Value As String

    Public Overrides ReadOnly Property DisplayValue As String
        Get
            Return Value
        End Get
    End Property

    Public Overrides ReadOnly Property IsDateRangeFilter As Boolean
        Get
            Return False
        End Get
    End Property

    Overrides ReadOnly Property Description As String
        Get
            Return $"{DisplayPropertyName} contiene {Value}"
        End Get
    End Property

    Sub New(propertyName As String, displayName As String, repo As IBussinessObjectRepository, value As String)
        MyBase.New(propertyName, displayName, repo)
        _Value = value
    End Sub

    Public Overrides Function Apply(originalList As IQueryable(Of T)) As IQueryable(Of T)
        Return From o In originalList
               Let s = CStr(CallByName(o, PropertyName, CallType.Get))
               Where Not String.IsNullOrEmpty(s) AndAlso s.ToLower.Contains(_Value.ToLower)
               Select o
    End Function

End Class

File: DesktopApp/src/Filters/StringFilter.cs
using Microsoft.VisualBasic;
using System.Linq;
using Wells.CorePersistence.Repositories;

namespace Wells.View.Filters
{
    public class StringFilter<T> : BaseFilter<T>
    {
        public string Value { get; set; }
        public override string DisplayValue => Value;

        public override bool IsDateRangeFilter => false;

        public override string Description
        {
            get
            {
                return $"{DisplayPropertyName} contiene {Value}";
            }
        }

        public StringFilter(string propertyName, string displayName, IBussinessObjectRepository repo, string value) :
            base(propertyName, displayName, repo)
        {
            Value = value;
        }

        public override IQueryable<T> Apply(IQueryable<T> queryable)
        {
            return from o in queryable
                   let s = (string)Interaction.CallByName(o, PropertyName, CallType.Get)
                   where !string.IsNullOrEmpty(s) && s.ToLower().Contains(Value.ToLower())
                   select o;
        }
    }
}
